File: Documentation/5.4/Samples/python/ClientApi/Session/Querying/FilterByNonExistingField.py
from ravendb import AbstractIndexCreationTask
from ravendb.infrastructure.orders import Order

from examples_base import ExampleBase


# region the_index
# Define a static index on the 'Orders' collection
# ================================================


class Orders_ByFreight(AbstractIndexCreationTask):
    class IndexEntry:
        def __init__(self, freight: int = None, Id: str = None):
            self.freight = freight
            self.Id = Id

    def __init__(self):
        # Call super().__init__() to initialize your index class
        super().__init__()
        # Define the index Map function
        self.map = "from o in docs.Orders select new { freight = o.freight, Id = o.Id }"


# endregion


class FilterByNonExistingField(ExampleBase):
    def setUp(self):
        super().setUp()
        with self.embedded_server.get_document_store("FilterByNonExistingField") as store:
            with store.open_session() as session:
                Orders_ByFreight().execute(store)
                self.add_orders(session)

    def test_filter_by_non_existing_field(self):
        with self.embedded_server.get_document_store("FilterByNonExistingField") as store:
            with store.open_session() as session:
                # region whereNotExists_1
                orders_without_freight_field = list(
                    session
                    # Define a DocumentQuery on 'Orders' collection
                    .document_query(object_type=Order)
                    # Search for documents that do not contain field 'freight'
                    .not_().where_exists("freight")
                )

                # Results will be only the documents that do not contain the 'freight' field in 'Orders' collection
                # endregion

            with store.open_session() as session:
                # region whereNotExists_2
                # Query the index
                # ===============
                fields = list(session.query_index_type(Orders_ByFreight, Orders_ByFreight.IndexEntry))
                orders_without_freight_field = list(
                    session
                    # Define a DocumentQuery on the index
                    .query_index_type(Orders_ByFreight, Orders_ByFreight.IndexEntry)
                    # Verify the index is not stale (optional)
                    .wait_for_non_stale_results()
                    # Search for documents that do not contain field 'freight'
                    .not_().where_exists("freight")
                )

                # Results will be only the documents that do not contain the 'freight' field in 'Orders' collection
                # endregion

File: Documentation/5.4/Samples/php/ClientApi/session/Querying/FilterByNonExistingField.php
<?php

use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Samples\Infrastructure\DocumentStoreHolder;

class FilterByNonExistingField
{
    public function examples(): void
    {
        $store = DocumentStoreHolder::getStore();
        try {
            $session = $store->openSession();
            try {
                # region whereNotExists_1
                /** @var array<Order> $ordersWithoutFreightField */
                $ordersWithoutFreightField = $session
                        ->advanced()
                         // Define a DocumentQuery on 'Orders' collection
                        ->documentQuery(Order::class)
                         // Search for documents that do Not contain field 'Freight'
                        ->not()
                        ->whereExists("Freight")
                         // Execute the query
                        ->toList();

                // Results will be only the documents that do Not contain the 'Freight' field in 'Orders' collection
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region whereNotexists_2
                // Query the index
                // ===============

                /** @var array<Order> $ordersWithoutFreightField */
                $ordersWithoutFreightField = $session
                    ->advanced()
                    // Define a DocumentQuery on the index
                    ->documentQuery(IndexEntry::class, Orders_ByFright::class)
                    // Verify the index is not stale (optional)
                    ->waitForNonStaleResults()
                    // Search for documents that do Not contain field 'Freight'
                    ->not()
                    ->whereExists("Freight")
                     // Execute the query
                    ->toList();

                // Results will be only the documents that do Not contain the 'Freight' field in 'Orders' collection
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

# region the_index
// Define a static index on the 'Orders' collection
// ================================================

class IndexEntry
{
    // Define the index-fields
    public ?float $freight = null;
    public ?string $id = null;

    public function getFreight(): float
    {
        return $this->freight;
    }

    public function setFreight(float $freight): void
    {
        $this->freight = $freight;
    }

    public function getId(): ?string
    {
        return $this->id;
    }

    public function setId(?string $id): void
    {
        $this->id = $id;
    }
}

class Orders_ByFright extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();
        // Define the index Map function
        $this->map = "orders => from doc in orders select new {\n" .
            "    freight = doc.name, \n" .
            "    id = doc.id\n" .
            "})";

        }
    }
# endregion
